File: Ch02/ex2_07.erl
-module(ex2_07).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Exercise 2.7-2.10 - Interval Arithmetic").
-export([add_interval/2, mult_interval/2, div_interval/2, make_interval/2, sub_interval/2, width/1]).
-include_lib("records.hrl").

make_interval(Lower, Upper) -> #interval{lower=Lower, upper=Upper}.

add_interval(I1, I2) -> make_interval(add_lower(I1, I2),
                                      add_upper(I1, I2)).

div_interval(I1, I2) ->
  case spans_zero(I1) of
    true  -> throw({badarith, "Can't divide when spanning zero."});
    false -> mult_interval(I1, recip(I2))
  end.

mult_interval(I1, I2) -> make_interval(lists:min(products(I1, I2)),
                                       lists:max(products(I1, I2))).

sub_interval(I1, I2)  -> make_interval(sub_lower(I1, I2),
                                       sub_upper(I1, I2)).

width(#interval{lower=Lower,upper=Upper}) -> (Upper - Lower) / 2.0.

% helper_methods
add_lower(#interval{lower=L1},
          #interval{lower=L2}) -> L1 + L2.

add_upper(#interval{upper=U1},
          #interval{upper=U2}) -> U1 + U2.

products(#interval{lower=L1,upper=U1},
         #interval{lower=L2,upper=U2}) ->
  [L1 * L2, L1 * U2, U1 * L2, U1 * U2].

recip(#interval{lower=Lower,upper=Upper}) ->
  make_interval((1.0 / Lower), (1.0 / Upper)).

spans_zero(#interval{lower=Lower,upper=Upper}) ->
  ((Lower >= 0) and (Upper =< 0)) or
  ((Lower =< 0) and (Upper >= 0)).

sub_lower(#interval{lower=L1},
          #interval{lower=L2}) -> L1 - L2.

sub_upper(#interval{upper=U1},
          #interval{upper=U2}) -> U1 - U2.

File: Ch02/ex2_07.exs
# SICP in Elixir: Exercise 2.7-2.10 - Interval Arithmetic

defmodule Interval do
  def new(lower, upper) do
    lower..upper
  end

  def add(l1..u1, l2..u2) do
    new(l1 + l2, u1 + u2)
  end

  def div(i1, i2) do
    if spans_zero?(i1) do
      throw({:badarg, "Can't divide when spanning zero."})
    else
      mult(i1, recip(i2))
    end
  end

  def mult(i1, i2) do
    new(:lists.min(products(i1, i2)), :lists.max(products(i1, i2)))
  end

  def sub(i1, i2) do
    new(sub_lower(i1, i2), sub_upper(i1, i2))
  end

  def width(lower..upper) do
    (upper - lower) / 2.0
  end

  # Private functions
  def products(l1..u1, l2..u2) do
    [l1 * l2, l1 * u2, u1 * l2, u1 * u2]
  end

  def recip(lower..upper) do
    new(1.0 / lower, 1.0 / upper)
  end

  defp spans_zero?(lower..upper) do
    (lower >= 0 and upper <= 0) or (lower <= 0 and upper >= 0)
  end

  defp sub_lower(l1.._, l2.._) do
    l1 - l2
  end

  defp sub_upper(_..u1, _..u2) do
    u1 - u2
  end
end

i1 = Interval.new(1, 5)
i2 = Interval.new(3, 4)
i3 = Interval.add(i1, i2)
i4 = Interval.div(i1, i2)
i5 = Interval.sub(i3, i4)
IO.puts(Interval.width(i5))
IO.puts(Interval.width(i1))
